lib/ja5_filter.h

```diff
@@ -0,0 +1,214 @@
+	TimeSlot		conns[JA5_FILTER_TIME_SLOTS_CNT];
+	spinlock_t		conns_lock;
+	TimeSlot		recs[JA5_FILTER_TIME_SLOTS_CNT];
+	spinlock_t		recs_lock;
+} Rates;
+
+
+static struct {
+	TDB			*tdb;
+	struct list_head	lru_list;
+	spinlock_t		lru_list_lock;
+} storage;
+
+static bool
+get_alloc_ctx_eq_rec(TdbRec *, void *)
+{
+	return true;
+}
+
+static void
+put_fingerprint_rates(Rates *rates)
+{
+	tdb_rec_put(storage.tdb, (char *)rates - sizeof(TdbRec));
+}
+
+static void
+get_alloc_ctx_init_rec(TdbRec *rec, void *)
+{
+	Rates *rates = (Rates *)rec->data;
+
+	bzero_fast(rates, sizeof(Rates));
+	INIT_LIST_HEAD(&rates->list_node);
+	spin_lock_init(&rates->conns_lock);
+	spin_lock_init(&rates->recs_lock);
+	tdb_rec_keep(rec);
+}
+
+static Rates*
+get_fingerprint_rates(u64 fingerprint)
+{
+	TdbGetAllocCtx get_alloc_ctx = {
+		.eq_rec = get_alloc_ctx_eq_rec,
+		.ctx = NULL,
+		.precreate_rec = NULL,
+		.init_rec = get_alloc_ctx_init_rec,
+		.len = sizeof(Rates)};
+	const u64 key = fingerprint;
+	TdbRec *rec;
+	Rates *rates;
+
+	/* Try to remove DB_RECS_TO_FREE_CNT records from DB if it's full */
+	while (!(rec = tdb_rec_get_alloc(storage.tdb, key, &get_alloc_ctx))) {
+		struct list_head *pos, *tmp, tail_to_delete;
+		u32 cnt = DB_RECS_TO_FREE_CNT;
+
+		INIT_LIST_HEAD(&tail_to_delete);
+
+		/* Cut off DB_RECS_TO_FREE_CNT entries from the LRU list */
+		spin_lock(&storage.lru_list_lock);
+		list_for_each_prev_safe(pos, tmp, &storage.lru_list) {
+			list_move(pos, &tail_to_delete);
+			if (!--cnt)
+				break;
+		}
+		spin_unlock(&storage.lru_list_lock);
+
+		list_for_each_safe(pos, tmp, &tail_to_delete)
+			put_fingerprint_rates((Rates *)pos);
+
+		/**
+		 * Protect from low probable case where all records
+		 * were deleted but are held by references
+		 */
+		if (cnt == DB_RECS_TO_FREE_CNT)
+			return NULL;
+	}
+
+	rates = (Rates *)rec->data;
+
+	spin_lock(&storage.lru_list_lock);
+	/* The record still was not added to the LRU list */
+	if (list_empty(&rates->list_node))
+		list_add_tail(&rates->list_node, &storage.lru_list);
+	spin_unlock(&storage.lru_list_lock);
+
+	return rates;
+}
+
+static bool
+init_filter(size_t max_storage_size)
+{
+	/**
+	 * Initialize storage only once during whole uptime.
+	 * Storage size reconfiguration is not supported.
+	 */
+	if (storage.tdb)
+		return true;
+
+	INIT_LIST_HEAD(&storage.lru_list);
+	spin_lock_init(&storage.lru_list_lock);
+
+	return (storage.tdb = tdb_open(
+		"/tmp/ja5t_flt.tdb", max_storage_size, sizeof(Rates), 0));
+}
+
+static u32
+ja5_calc_rate(TimeSlot slots[])
+{
+	u32 sum = 0;
+	u64 ts = jiffies * JA5_FILTER_TIME_SLOTS_CNT / HZ;
+	u64 end_ts = ts - JA5_FILTER_TIME_SLOTS_CNT;
+	u8 slot_num = ts % JA5_FILTER_TIME_SLOTS_CNT;
+	TimeSlot *slot = &slots[slot_num];
+
+	if (slot->ts != ts) {
+		slot->ts = ts;
+		slot->counter = 0;
+	}
+	slot->counter++;
+
+	for (; slot->ts > end_ts;
+		slot_num = (slot->ts - 1) % JA5_FILTER_TIME_SLOTS_CNT,
+		slot = &slots[slot_num])
+		sum += slot->counter;
+
+	return sum;
+}
```

Comment:
The computation of the rate limit under spin lock doesn't look too advances. However, at the moment we use somewhat similar rate limit in the XDP volumetric DDoS mitigation module and it easily process more than 100Gbps on adequate CPU power.

Scalable and efficient rate limits are hard. There is [Hierarchical Heavy Hitters with the Space Saving Algorithm](https://arxiv.org/pdf/1102.5540.pdf) proposes probabilistic space efficient approach, which is said is easy to be parallelized and merged. We also discussed per-cpu hierarchical approach.

I think we firstly need to solve the problem on the XDP  (and more performance crucial) level and next in L7 (Tempesta FW).